x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -419,7 +420,13 @@ protected LogicalPlan rule(UnaryPlan plan) {
             // Agg with underlying Project (group by on sub-queries)
             if (plan instanceof Aggregate a) {
                 if (child instanceof Project p) {
-                    plan = new Aggregate(a.source(), p.child(), a.groupings(), combineProjections(a.aggregates(), p.projections()));
+                    plan = new Aggregate(
+                        a.source(),
+                        p.child(),
+                        // After substitutions groupings can only contain attributes.
+                        combineUpperGroupingsAndLowerProjections((List<? extends Attribute>) (List<?>) a.groupings(), p.projections()),
```

Comment:
```suggestion
                        combineUpperGroupingsAndLowerProjections((List<? extends Attribute>) a.groupings(), p.projections()),
```

Reply:
Ah, sorry, this suggestion is wrong. 
Why the double cast though and not call the combining function with a `List<Expression>` and cast downstream, since groupings are expected as attributes? Same cast as for the projection list.

Reply:
Nevermind, explicitly casting each grouping now per Andrei's suggestion.